salt-mode.el

```diff
@@ -194,6 +213,202 @@ module to use.
          (method (cadr pieces)))
     (browse-url (if method (concat url "#salt.states." module "." method) url))))
 
+(defconst salt-mode--query-template "
+try:
+    import salt.config
+    from salt.minion import SMinion
+    import json
+    opts = salt.config.minion_config('')
+    opts['file_client'] = 'local'
+    minion = SMinion(opts)
+except Exception as exc:
+    raise SystemExit(repr(exc))
+else:
+    print(json.dumps(%s, indent=4))"
+  "Python template to query the Salt minion state.
+
+This does not load the full local minion configuration, so will not
+be able to access custom states and modules.")
+
+(defun salt-mode--query-minion (program)
+  "Run Python code PROGRAM on a virtual Salt minion.
+
+Error handling is the responsibility of the caller; failures
+may occur starting the Python process or parsing its output."
+  (with-temp-buffer
+    (process-file salt-mode-python-program nil t nil "-c"
+                  (format salt-mode--query-template program))
+    (goto-char (point-min))
+    (let ((json-array-type 'list)
+          (json-object-type 'hash-table))
+      (json-read))))
+
+(defun salt-mode--async-minion (program callback)
+  "Run Python code PROGRAM on a virtual Salt minion.
+
+JSON response data is passed to CALLBACK when it is ready. Errors
+result in a message but no signal."
+  (with-demoted-errors "Unable to query Salt minion: %s"
+    (message "starting in %s" default-directory)
+    (set-process-sentinel
+     (start-file-process "salt-async"
+                         ;; Work around TRAMP bug in TRAMP 2.2 / Emacs 25.1;
+                         ;; remote buffers cannot be passed by name.
+                         (get-buffer-create
+                          (generate-new-buffer-name " *salt-async*"))
+                         salt-mode-python-program "-c"
+                         (format salt-mode--query-template program))
+     (lambda (process _event)
+       (when (memq (process-status process) '(exit signal))
+         (with-current-buffer (process-buffer process)
+           (with-demoted-errors "Error querying Salt minion: %s"
+             (goto-char (point-min))
+             (condition-case nil
+                 (let ((json-array-type 'list)
+                       (json-object-type 'hash-table))
+                   (funcall callback (json-read)))
+               ((json-readtable-error)
+                (message "Error querying Salt minion: %s"
+                         (string-trim-right (buffer-string))))))
+           (kill-buffer)))))))
+
+(defmacro salt-mode--with-async-minion (program &rest body)
+  "Run Python code PROGRAM on a virtual Salt minion.
+
+BODY will run once the result is available, in the variable `result'."
+  (declare (indent 1))
+  `(salt-mode--async-minion ,program (lambda (result) ,@body)))
+
+
+(defvar salt-mode--state-argspecs nil
+  "Information about Salt states.")
+
+(defun salt-mode-refresh-data (&optional if-missing)
```

Comment:
Whoops, "starting in..." is a debug print from trying to work around that TRAMP bug. That will be removed.

Rather than silencing it completely, I'd prefer to inhibit in the echo area but leave it in the message buffer. I have a feeling someone seeing "Loaded 0 Salt states" will save a lot of time debugging something in the future.